### PR TITLE
Mock Service Worker Conversion - Initial Setup

### PR DESCRIPTION
**Pull request by JacobMGEvans**
Initial setup for replacing custom mocking and fetching layer mocking with
Mock Service Worker allowing for mocking in the network layer of requests and responses,
which will allow for full testing of code with minimal mocking of it.

Tests converted to Mock Service Worker w/ included request input checks in handlers inline in certain tests:
- r2.test.ts
- whoami.test.ts

### CODE REVIEW

**Comment by JacobMGEvans** [comment]
Special thanks to @caass for the fetch fall-through logic

**Comment by JacobMGEvans** [comment]
emphasizing the Delta on internal 😆 
image

File: packages/wrangler/src/__tests__/r2.test.ts [hunk 1]
```diff
@@ -148,32 +141,35 @@ describe("wrangler", () => {
...
-				it("should create a bucket", async () => {
-					const requests = mockCreateRequest("testBucket");
+				it("should create a bucket & check request inputs", async () => {
+					msw.use(
+						rest.post(
+							"*/accounts/:accountId/r2/buckets",
+							async (request, response, context) => {
+								const { accountId } = request.params;
+								expect(accountId).toEqual("some-account-id");
+								expect(await request.json()).toEqual({ name: "testBucket" });
```
Comment (Aprillion) [inline packages/wrangler/src/__tests__/r2.test.ts:151]: the `"testBucket"` assertion is related to `runWrangler("r2 bucket create testBucket")` from this test, so that would be easy to investigate a potential failure ... but where does `"some-account-id"` come from? if that assertion failed, would it be easy to investigate that assertion?

**Review by Aprillion — APPROVED** [review]
Looks good to me in general, just a few comments.

File: packages/wrangler/src/__tests__/helpers/mock-cfetch.ts [hunk 2]
```diff
@@ -221,21 +239,24 @@ export async function mockFetchR2Objects(
...
+			bodyInit.body.removeAllListeners();
+		}
+
 		const value = r2GetMocks.get(resource);
 
 		return new Response(value);
 	}
-	throw new Error(`no mock found for \`r2 object\` - ${resource}`);
+
+	// No mocks found for ${init.method ?? "any HTTP"} request to ${resource}
```
[the remark this answers is not in the document]
Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/helpers/mock-cfetch.ts:257]: Its relevant but I will remove it anyway.

File: packages/wrangler/src/__tests__/helpers/msw/handlers/user.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+						created_by: "1fc1df98cc4420fe00367c3ab68c1639",
+						modified_on: "2022-06-29T14:30:08.16152Z",
+						modified_by: "1fc1df98cc4420fe00367c3ab68c1639",
+					},
+				})
+			);
+		}
+	),
+	rest.get(
+		"*/accounts/:accountId/workers/dispatch/namespaces",
```
[the remark this answers is not in the document]
Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/helpers/msw/handlers/user.ts:121]: I will move the namespaces ones into their own file. They aren't related directly to `user`

File: packages/wrangler/src/__tests__/r2.test.ts [hunk 4]
```diff
@@ -50,26 +43,41 @@ describe("wrangler", () => {
...
+						{ name: "bucket-2-local-once", creation_date: "01-01-2001" },
 					];
-					mockListRequest(expectedBuckets);
+					msw.use(
+						rest.get(
+							"*/accounts/:accountId/r2/buckets",
+							async (request, response, context) => {
+								const { accountId } = request.params;
+								expect(accountId).toEqual("some-account-id");
+								expect(await request.text()).toEqual("");
```
[the remark this answers is not in the document]
Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/r2.test.ts:57]: Because that was what the previous custom set mock logic was doing.

File: packages/wrangler/src/__tests__/r2.test.ts (hunk 1 above)
Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/r2.test.ts:151]: Yes, that is part of the initial account setup for tests. It happens in one place for all tests. 
https://github.com/cloudflare/wrangler2/pull/1768/files#diff-129fe75559da917c336722e1973c0294dde22f66b3affceb26cd95ad93c4c94bL14

File: packages/wrangler/package.json [hunk 5]
```diff
@@ -98,6 +98,7 @@
 		"chokidar": "^3.5.3",
 		"esbuild": "0.14.51",
 		"miniflare": "^2.8.1",
+		"msw": "^0.46.1",
```
Comment (petebacondarwin) [inline packages/wrangler/package.json:101]: Definitely should be a devDep not a dep.

File: packages/wrangler/src/__tests__/helpers/mock-cfetch.ts [hunk 6]
```diff
@@ -43,12 +59,14 @@ export async function mockFetchInternal(
...
-			return await handler(uri, init, queryParams); // TODO: should we have some kind of fallthrough system? we'll see.
+			return await handler(uri, init, queryParams ?? new URLSearchParams()); // TODO: should we have some kind of fallthrough system? we'll see.
 		}
 	}
-	throw new Error(
-		`no mocks found for ${init.method ?? "any HTTP"} request to ${resource}`
-	);
+	// no mocks found for ${init.method ?? "any HTTP"} request to ${resource}
+	// let it fall through to mock-service-worker
+	// (do a real, unmocked fetch)
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/helpers/mock-cfetch.ts:67]: Is this line correct? It is not doing a real unmocked fetch, right? All fetches, one way or another, should be mocked in tests.

File: packages/wrangler/package.json (hunk 5 above)
Comment (petebacondarwin) [reply packages/wrangler/package.json:101]: Make sure you re-run `npm i` after fixing to ensure that package-lock.json gets updated.

File: packages/wrangler/src/__tests__/helpers/msw/handlers/oauth.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+				messages: [],
+				code: "test-oauth-code",
+			})
+		);
+	}),
+	// revoke access token
+	rest.post(
+		"https://dash.cloudflare.com/oauth2/revoke",
+		(_, response, context) => {
+			response(context.status(200), context.text(""));
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/wrangler/src/__tests__/helpers/msw/handlers/oauth.ts]: This callback is still returning void. Is that wanted?

**Review by petebacondarwin — COMMENTED** [review]
I feel a bit uncomfortable about how we mock out a load of "happy path" requests by default.
Although this might make writing tests simpler, it means that it is not possible to tell in any given tests, which endpoints we expect to hit and which not.
Have you considered using helpers in specific tests to add in handlers to msw on a needs-must basis?

**Comment by JacobMGEvans** [comment]
> I feel a bit uncomfortable about how we mock out a load of "happy path" requests by default. Although this might make writing tests simpler, it means that it is not possible to tell in any given tests, which endpoints we expect to hit and which not. Have you considered using helpers in specific tests to add in handlers to msw on a needs-must basis?

The pattern is that the test without inlined Mocks would be using the default handlers, the inline mocks which look like this `msw.use(rest.<method>(URL, (req, res) => response.once())` would indicate a deviation from the default behavior, explicitly indicating what paths are being tested.

> Have you considered using helpers in specific tests to add in handlers to msw on a needs-must basis?

This would basically be using the inline handler pattern (even if abstracted with a helper) for every test or test suite not just edge cases or deviations from "happy path". I feel like this misses out on a MSW standard/practice of top level handlers with the initialized MSW server.

File: packages/wrangler/package.json (hunk 5 above)
Comment (JacobMGEvans) [reply packages/wrangler/package.json]: Done 😄

**Review by caass — APPROVED** [review]
This looks fine to me -- great work! A few thoughts for future work:

1. It feels like there should be one level of abstraction here. I'm not sure what, exactly, yet, but I think it will become obvious as we write more of these tests.
2. It's worth creating a tracking issue for the tests that still have to be converted, and smaller issues tracked in there.
3. Some small helpers (e.g. `ctx.ok` for JSON ok responses and `ctx.error` for JSON error responses) wouldn't go amiss, I don't think

**Comment by JacobMGEvans** [comment]
> This looks fine to me -- great work! A few thoughts for future work:
> 
> 1. It feels like there should be one level of abstraction here. I'm not sure what, exactly, yet, but I think it will become obvious as we write more of these tests.
> 2. It's worth creating a tracking issue for the tests that still have to be converted, and smaller issues tracked in there.
> 3. Some small helpers (e.g. `ctx.ok` for JSON ok responses and `ctx.error` for JSON error responses) wouldn't go amiss, I don't think

I agree with these 3 points and will create an issue regarding them, then break them into separate issues

File: packages/wrangler/src/__tests__/helpers/mock-cfetch.ts (hunk 6 above)
Comment (petebacondarwin) [reply packages/wrangler/src/__tests__/helpers/mock-cfetch.ts:67]: That is what I mean. On the line above you say "fall through to the mock-service-worker" but on this line you say "do a real unmocked fetch". The fetch is still mocked, right? But by the MSW.
Perhaps this line could read, "(do a real, unmocked cfetch)"?

Comment (Aprillion) [reply packages/wrangler/src/__tests__/helpers/mock-cfetch.ts:67]: the actual unharmed global `fetch` (or the one from fetch-node) is really used and MSW only extends the `end` method of `class NodeClientRequest extends ClientRequest`, see https://github.com/mswjs/interceptors#readme

Comment (petebacondarwin) [reply packages/wrangler/src/__tests__/helpers/mock-cfetch.ts:67]: Ah thanks for the clarification. In that case, I would just remove this line. It doesn't add any value, given that we have already stated we are falling through to the MSW.
